### PR TITLE
cxx-qt-gen: move to syn 2

### PR DESCRIPTION
**Pull request by ahayzen-kdab**

### CODE REVIEW

File: crates/cxx-qt-gen/src/syntax/types.rs [hunk 1]
```diff
@@ -9,31 +9,56 @@ use syn::{
...
+    if let Type::Path(TypePath { path, .. }) = ty {
+        if path_compare_str(path, &["Pin"]) {
+            return Some(path.clone());
+        }
+    }
+
+    None
+}
+
+pub fn is_pin_mut(ty: &Type) -> bool {
```
Comment (ahayzen-kdab) [inline crates/cxx-qt-gen/src/syntax/types.rs:22]: need tests

File: crates/cxx-qt-gen/src/syntax/implitemfn.rs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+use syn::{FnArg, Receiver, Signature};
+
+use super::types::is_pin_mut;
+
+/// Return whether the first parameter of a method is a mutable self argument
+pub fn is_method_mutable(signature: &Signature) -> bool {
+    match signature.inputs.first() {
+        Some(FnArg::Receiver(Receiver { mutability, ty, .. })) => {
+            // Check if Pin<T> is mut or fallback to normal mutability
+            is_pin_mut(ty) || mutability.is_some()
```
Comment (LeonMatthesKDAB) [inline crates/cxx-qt-gen/src/syntax/implitemfn.rs:15]: We could use this change to fix some of the TODO's in the earlier code by changing this code to also check whether the receiver is a pin of self:
```suggestion
            (is_pin_of_self(ty) && is_pin_mut(ty)) || mutability.is_some()
```

Comment (ahayzen-kdab) [reply crates/cxx-qt-gen/src/syntax/implitemfn.rs:15]: Ah an it is using `Receiver` to find if it's a `self`

File: crates/cxx-qt-gen/src/syntax/foreignmod.rs [hunk 3]
```diff
@@ -112,50 +113,43 @@ fn verbatim_to_foreign_type(tokens: &TokenStream) -> Result<Option<ForeignItemTy
...
+            return Err(Error::new(
+                receiver.span(),
+                "mut on the `self:` receiver are not supported",
+            ));
         }
+
+        if receiver.reference.is_some() {
+            return Err(Error::new(
+                receiver.span(),
+                "Reference on the `self:` receiver are not supported",
```
Comment (LeonMatthesKDAB) [inline crates/cxx-qt-gen/src/syntax/foreignmod.rs:135]:
```suggestion
                "References on self (i.e. `&self`) are not supported, use `self: &T` instead!",
```

File: crates/cxx-qt-gen/src/syntax/foreignmod.rs [hunk 4]
```diff
@@ -112,50 +113,43 @@ fn verbatim_to_foreign_type(tokens: &TokenStream) -> Result<Option<ForeignItemTy
...
+            ));
+        }
+
+        if receiver.colon_token.is_none() {
+            return Err(Error::new(
+                receiver.span(),
+                "Missing type on the `self:` receiver are not supported",
```
Comment (LeonMatthesKDAB) [inline crates/cxx-qt-gen/src/syntax/foreignmod.rs:142]:
```suggestion
                "`self` is not supported as receiver, use `self: T` to indicate a type.",
```

File: crates/cxx-qt-gen/src/syntax/foreignmod.rs [hunk 5]
```diff
@@ -112,50 +113,43 @@ fn verbatim_to_foreign_type(tokens: &TokenStream) -> Result<Option<ForeignItemTy
...
-                    return Ok(ForeignFnSelf {
-                        ident: ident.clone(),
-                        typ: *ty.clone(),
-                    });
-                }
-            }
+        if receiver.mutability.is_some() {
+            return Err(Error::new(
+                receiver.span(),
+                "mut on the `self:` receiver are not supported",
```
Comment (LeonMatthesKDAB) [inline crates/cxx-qt-gen/src/syntax/foreignmod.rs:128]:
```suggestion
                "mut on the `self:` receiver is not supported",
```

Comment (ahayzen-kdab) [reply crates/cxx-qt-gen/src/syntax/foreignmod.rs:128]: Shouldn't this be "mut on self (i.e. `&mut self`) are not supported, use `self: Pin<&mut T>` instead" ?

Comment (LeonMatthesKDAB) [reply crates/cxx-qt-gen/src/syntax/foreignmod.rs:128]: Ah, yes, good point :+1:

File: crates/cxx-qt-gen/src/syntax/implitemfn.rs (hunk 2 above)
Comment (ahayzen-kdab) [reply crates/cxx-qt-gen/src/syntax/implitemfn.rs]: Done

File: crates/cxx-qt-gen/src/syntax/foreignmod.rs (hunk 3 above)
Comment (ahayzen-kdab) [reply crates/cxx-qt-gen/src/syntax/foreignmod.rs]: Done

File: crates/cxx-qt-gen/src/syntax/foreignmod.rs (hunk 4 above)
Comment (ahayzen-kdab) [reply crates/cxx-qt-gen/src/syntax/foreignmod.rs]: Done

File: crates/cxx-qt-gen/src/syntax/foreignmod.rs (hunk 5 above)
Comment (ahayzen-kdab) [reply crates/cxx-qt-gen/src/syntax/foreignmod.rs]: Done

**Review by LeonMatthesKDAB — APPROVED** [review]
LGTM

**Pull request merged**